src/cpp/src/speculative_decoding/speculative_decoding_impl.cpp

```diff
@@ -257,6 +259,52 @@ ContinuousBatchingPipeline::SpeculativeDecodingImpl::generate(const std::vector<
                     break;
                 }
             }
+
+            // todo: remove
+            streaming_timer.end();
+            streaming_duraton += streaming_timer.get_duration();
+        }
+    };
+
+    // to store potential exception thrown in step_thread
+    std::exception_ptr step_outputs_error = nullptr;
+    while (continue_generation) {
+        // todo: remove
+        ManualTimer thread_timer("threading");
+        thread_timer.start();
+
+        // to define inference thread
+        std::thread t_step([this, &step_outputs_error] {
+            try {
+                step();
+            } catch (...) {
+                // remove all requests from pipeline state in case of exception
+                drop_requests();
+                step_outputs_error = std::current_exception();
+            }
+        });
+
+        // to define streaming thread
+        std::thread t_stream([&stream_generated_tokens] {
```

Comment:
I think we need to create a thread before this `while` loop:
```c++
std::thread ([] { 
   while (can_read()) { // uses thread safe queue
    read and stream
  }
})
```
and join this thread after this `while`

Reply:
Implemented